### ISSUE TITLE
App upgrade overwrites configuration file

### ISSUE THREAD

**Issue by weex** [YunoHost/issues#1948]
### Describe the bug

When upgrading Ecko the configuration file, which admittedly is stored with the code itself, gets overwritten by the processed default version.

### Context

- Hardware: vps
- YunoHost version: 4.3.4.1
- I have access to my server: Through SSH && through the webadmin && direct access via keyboard / screen
- Are you in a special context or did you perform some particular tweaking on your YunoHost instance?: no
- Using, or trying to install package version/branch: Ecko 2021-12-09~ynh1
- If upgrading, current YunoHost version: n/a

### To reproduce

1. Install Ecko (any previous version)
2. Make configuration changes in `/var/www/ecko/live/.env.production`
3. Upgrade Ecko

### Expected behavior

Configuration file isn't overwritten. Instead of existing version being backed up, it would be better for this app for it to stay in place and the newly generated version placed in an alternate location should it need to be referenced.

### Logs

https://paste.yunohost.org/raw/pilalilovo

**Comment by tituspijean** [comment]
Refer to `ynh_setup_source`'s helper documentation: https://yunohost.org/oc/packaging_apps_helpers#ynh_setup_source

It has a `--keep` just for this case.

Another option is to store all relevant parameters within YunoHost and use `ynh_add_config` to regenerate it properly. But it's more complex and tiresome to maintain.

**Comment by weex** [comment]
From @tituspijean: (I got a notification that this was here but didn't see it so posting)

Refer to ynh_setup_source's helper documentation: https://yunohost.org/oc/packaging_apps_helpers#ynh_setup_source

It has a --keep just for this case.

Another option is to store all relevant parameters within YunoHost and use ynh_add_config to regenerate it properly. But it's more complex and tiresome to maintain.

**Comment by tituspijean** [comment]
Yeah sorry, I deleted my comment because, after reading it again, it seemed it was unrelated to your issue. 😅

**Comment by weex** [comment]
Any reason? I haven't tried it but was to be next when I get back to fixing it in the linked issue.

**Comment by alexAubin** [comment]
>Configuration file isn't overwritten. Instead of existing version being backed up, it would be better for this app for it to stay in place and the newly generated version placed in an alternate location should it need to be referenced.

It's a complicated issue : 
- for many apps, the format of the configuration evolves and it's just simpler to overwrite the existing conf, because anyway in the context of Yunohost, it's expected that people won't edit that file manually, and you don't want to have to tell them to manually cp some new conf file. Both because people are not supposed to be tech-savy enough to do so, and because a reasonably designed app will expose the proper options in its web ui.
- for some other apps meant for more tech-savvy people though, it's somewhat expected that people will manually edit the conf file. In this case, the standard behavior is "overwrite the conf file, but if the file was found to be manually modified, display a warning and save the current conf somewhere and warn the admin about it". Currently that warning is somewhat burried in the log and you may easily miss it i guess, but with the new packaging format, we could trigger a proper popup modal to warn about it.

Altenatively, the third option to be an intermediate is to expose all settings which the admin is likely to be willing to tweak through an app config panel inside Yunohost's webadmin

**Comment by eauchat** [comment]
Just to add an additional bit of context.
I've encountered the same issue with etherpad, having customized the theme, I have to reapply the customization from scratch every time I update etherpad (= retweak configuration file, but also every theme files).
I don't know if there's any simple answer to the issue, because for example in that case, the theme consists of tens of css, js files, images... Identifying which files to keep and which to change on update doesn't seem so doable.

In such context, I don't know how easy it is to track so many files for manual changes, but yes if that's doable, having a warning about it would definitely be a plus (because the first time it happened, in my etherpad example, I didn't expect it, and I had to reinvent the theme because it was just lost ;)).

IMHO not overwriting the modified configuration would probably be a bad design as it would end up breaking apps in many cases. But if there's a way, not too complicated to implement, to warn the user that their changes will be lost, that's definitely a plus.
